### PR TITLE
Separate plugs and execution

### PR DESCRIPTION
**Pull request by fahhem**
Moves the plug manager creation into early test creation and makes its API more concrete.

I also separated the output code into `Test.Execute()` which cleaned up that separation a little, too. I'm not attached to that if either of you don't want that.

### CODE REVIEW

File: openhtf/__init__.py [hunk 1]
```diff
@@ -103,8 +103,13 @@ def AddOutputCallbacks(self, *callbacks):
     """Add the given function as an output module to this test."""
     self._test_options.output_callbacks.extend(callbacks)
 
-  def OutputTestRecord(self, record):
+  def OutputTestRecord(self):
     """Feed the record of this test to all output modules."""
+    test_state = self._executor.GetState()
+    if not test_state:
+      return
```
Comment (grybmadsci) [inline openhtf/__init__.py]: this code path probably warrants a log message

File: openhtf/__init__.py [hunk 2]
```diff
@@ -103,8 +103,13 @@ def AddOutputCallbacks(self, *callbacks):
...
+
+    record = test_state.GetFinishedRecord()
     for output_cb in self._test_options.output_callbacks:
       output_cb(record)
```
Comment (grybmadsci) [inline openhtf/__init__.py]: while you're here, can you add a `try:` block around each callback so exceptions in output callbacks don't cause this thread to die?

File: openhtf/exe/phase_data.py [hunk 3]
```diff
@@ -72,9 +72,7 @@ def _asdict(self):
     """Return a dict of this PhaseData's public data."""
     return {'measurements': self.measurements,
             'attachments': self.attachments.keys(),
-            'plugs': {
-                k.__module__ + '.' + k.__name__: str(v)
-                for k, v in self.plug_map.iteritems()}}
+            'plug_manager': self.plug_manager}
```
Comment (grybmadsci) [inline openhtf/exe/phase_data.py]: hm, i forget why this even needed an `_asdict` in the first place.  if possible, we should get rid of this.  if it's needed, then I'm guessing it's somewhere that expects all base types, so i'd rather keep the `plugs` dict and access some helper method on `self.plug_manager` to pull the strings out that we used to generate directly on `self.plug_map`

File: openhtf/plugs/__init__.py [hunk 4]
```diff
@@ -184,19 +185,29 @@ class PlugManager(object):
...
       for plug_type in plug_types:
-        self.plug_map[plug_type] = plug_type()
+        self._plug_map[plug_type] = plug_type()
     except Exception:  # pylint: disable=broad-except
       _LOG.exception('Exception insantiating plug type %s', plug_type)
       self.TearDownPlugs()
       raise
 
+  @property
+  def _asdict(self):
```
Comment (grybmadsci) [inline openhtf/plugs/__init__.py]: ah, okay, i see what you did, you're relying on the recursive nature of `ConvertToBaseTypes`.. that's fine actually, but can you look into where that's actually called, and potentially just remove both `_asdict()` methods if they're not used?  they might be used for the http_api, I think...

File: openhtf/exe/__init__.py [hunk 5]
```diff
@@ -67,14 +66,12 @@ class TestExecutor(threads.KillableThread):
                          ['CREATED', 'START_WAIT', 'INITIALIZING', 'EXECUTING',
                           'STOP_WAIT', 'FINISHING'])
 
-  def __init__(self, test):
+  def __init__(self, test_data, plug_manager):
     super(TestExecutor, self).__init__(name='TestExecutorThread')
 
-    self._test = test
+    self._test_data = test_data
+    self._plug_mgr = plug_manager
```
Comment (grybmadsci) [inline openhtf/exe/__init__.py]: can we call it `self._plug_manager` here to be consistent?

**Comment by grybmadsci** [comment]
LGTM, pending minor nits

File: openhtf/__init__.py (hunk 1 above)
Comment (fahhem) [inline openhtf/__init__.py]: I changed to take the finished record again, I don't know why I took that out.

File: openhtf/__init__.py (hunk 2 above)
Comment (fahhem) [inline openhtf/__init__.py]: Right, good idea. Wasn't there a bug for this? I can't find it.

File: openhtf/exe/__init__.py (hunk 5 above)
Comment (fahhem) [inline openhtf/exe/__init__.py]: Done

File: openhtf/plugs/__init__.py (hunk 4 above)
Comment (fahhem) [inline openhtf/plugs/__init__.py]: it's included in TestState->PhaseData. The templates refer to measurements inside of PhaseData, but not the plugs, so I could remove them for now. Though I'm pretty sure Joe's frontend would use them

**Comment by fahhem** [comment]
nits addressed

**Pull request merged**